File: upac-cli/src/commands/rollback.rs
// ── Imports ─────────────────────────────────────────────────────────────────
use anyhow::Result;

use indicatif::ProgressBar;

use colored::Colorize;

use std::ffi::CString;
use std::sync::Arc;

use crate::cancel_token_ptr;
use crate::config::Config;
use crate::ffi::CMutatedRequest;
use crate::upac::UpacLib;
use crate::utils::spinner;

// ── Arguments for command ───────────────────────────────────────────────────────────────────────
#[derive(clap::Args)]
pub struct RollbackArgs {
    pub commit: CString,
}

// ── FSM states ───────────────────────────────────────────────────────────────────────
#[derive(Debug, Clone, PartialEq)]
enum State {
    Validating,
    RollingBack,
    Done,
}

// ── FSM machine ───────────────────────────────────────────────────────────────────────
struct RollbackMachine {
    commit_hash: CString,

    upac_lib: Arc<UpacLib>,
    progress_bar: ProgressBar,
    config: Config,
    state: State,
}

impl RollbackMachine {
    fn new(config: Config, upac_lib: Arc<UpacLib>, commit_hash: CString) -> Result<Self> {
        Ok(Self {
            commit_hash,
            progress_bar: ProgressBar::new_spinner(),
            upac_lib: upac_lib,
            config,
            state: State::Validating,
        })
    }
}

// ── Public API ─────────────────────────────────────────────────────────────
pub fn run(args: RollbackArgs, config: Config, upac_lib: Arc<UpacLib>) -> Result<()> {
    let mut rolling_machine = RollbackMachine::new(config, upac_lib, args.commit)?;

    state_validating(&mut rolling_machine).map_err(|err| {
        if rolling_machine.config.verbose {
            eprintln!(
                "{} failed at state {:?}",
                "✗".red().bold(),
                rolling_machine.state
            );
        }
        err
    })
}

// ── States ─────────────────────────────────────────────────────────────────
fn state_validating(machine: &mut RollbackMachine) -> Result<()> {
    machine.state = State::Validating;
    spinner(&machine.progress_bar, "Validating rolling data...");

    if machine.commit_hash.as_bytes().len() != 64
        || !machine
            .commit_hash
            .as_bytes()
            .iter()
            .all(|byte| byte.is_ascii_hexdigit())
    {
        anyhow::bail!(
            "invalid commit hash '{:?}'. Expected 64 hex characters",
            machine.commit_hash
        );
    }

    machine.progress_bar.println(format!(
        "{} rolling back to {}",
        "→".cyan(),
        &machine.commit_hash.to_str()?[..12].dimmed()
    ));

    state_rolling_back(machine)
}

fn state_rolling_back(machine: &mut RollbackMachine) -> Result<()> {
    machine.state = State::RollingBack;
    spinner(&machine.progress_bar, "Rolling back...");

    let token_ptr = cancel_token_ptr();

    let rollback_request_c = CMutatedRequest::for_rollback(
        &machine.commit_hash,
        &machine.config.paths.repo_path,
        &machine.config.paths.root_path,
        &machine.config.paths.database_path,
        &machine.config.ostree.branch,
        &machine.config.ostree.prefix_directory,
        machine.config.step_retries,
        token_ptr,
    );

    UpacLib::check(
        unsafe { (machine.upac_lib.as_ref().rollback)(rollback_request_c) },
        "rollback",
    )?;

    state_done(machine)
}

fn state_done(machine: &mut RollbackMachine) -> Result<()> {
    machine.state = State::Done;
    machine.progress_bar.finish_and_clear();

    println!(
        "{} rolled back to {}",
        "✓".green().bold(),
        &machine.commit_hash.to_str()?[..12].bold()
    );

    Ok(())
}

File: lib/src/symbols/mutated/rollback.zig
// ── Imports ─────────────────────────────────────────────────────────────────────
const std = @import("std");

const ffi = @import("upac-ffi");

const CRollbackRequest = ffi.CMutatedRequest;

const types = @import("upac-types");

const ErrorCode = types.ErrorCode;
const Operation = types.Operation;
const fromError = types.fromError;

const rollback_module = @import("upac-rollback");
const RollbackData = rollback_module.RollbackData;
const RollbackMachine = rollback_module.RollbackMachine;

// Reverts the system state to a specific commit hash in the OSTree repository
pub fn rollback(rollback_request_c: CRollbackRequest) callconv(.c) i32 {
    rollback_request_c.validate() catch |err| return @intFromEnum(fromError(err, Operation.rollback));

    if (rollback_request_c.commit_hash.len == 0) return @intFromEnum(fromError(error.InvalidEntry, Operation.rollback));
    rollback_request_c.commit_hash.validate() catch return @intFromEnum(fromError(error.InvalidEntry, Operation.rollback));

    const rollback_data = RollbackData{
        .root_path = rollback_request_c.root_path.asZ(),
        .repo_path = rollback_request_c.repo_path.asZ(),
        .branch = rollback_request_c.branch.asZ(),
        .commit_hash = rollback_request_c.commit_hash.asZ(),
        .cancel_token = rollback_request_c.cancel_token orelse return @intFromEnum(fromError(error.InvalidEntry, Operation.rollback)),
    };

    RollbackMachine.run(rollback_data, ffi.getAllocator()) catch |err| return @intFromEnum(fromError(err, Operation.rollback));

    return @intFromEnum(ErrorCode.ok);
}
